vendor/github.com/mongodb/mongo-tools-common/db/buffered_bulk.go

```diff
@@ -12,85 +12,98 @@ import (
+		docLimit:      docLimit,
+		writeModels:   make([]mongo.WriteModel, 0, docLimit),
 	}
-	bb.resetBulk()
 	return bb
 }
 
 func (bb *BufferedBulkInserter) Unordered() {
-	bb.unordered = true
+	bb.bulkWriteOpts.SetOrdered(false)
 }
 
 func (bb *BufferedBulkInserter) SetBypassDocumentValidation(bypass bool) {
-	bb.bypassDocumentValidation = bypass
+	bb.bulkWriteOpts.SetBypassDocumentValidation(bypass)
+}
+
+func (bb *BufferedBulkInserter) SetUpsert(upsert bool) {
+	bb.upsert = upsert
 }
 
 // throw away the old bulk and init a new one
 func (bb *BufferedBulkInserter) resetBulk() {
-	bb.documents = make([]interface{}, 0, bb.docLimit)
-	bb.byteCount = 0
+	bb.writeModels = bb.writeModels[:0]
 	bb.docCount = 0
 }
 
-// Insert adds a document to the buffer for bulk insertion. If the buffer is
-// full, the bulk insert is made, returning any error that occurs.
+// Insert adds a document to the buffer for bulk insertion. If the buffer becomes full, the bulk write is performed, returning
+// any error that occurs.
 func (bb *BufferedBulkInserter) Insert(doc interface{}) error {
 	rawBytes, err := bson.Marshal(doc)
 	if err != nil {
-		return fmt.Errorf("bson encoding error: %v", err)
+		return fmt.Errorf("bson encoding err: %v", err)
 	}
 
 	return bb.InsertRaw(rawBytes)
 }
 
-// InsertRaw adds a document, represented as raw bson bytes, to the buffer for bulk insertion. If the buffer is full,
-// the bulk insert is made, returning any error that occurs.
-func (bb *BufferedBulkInserter) InsertRaw(rawBytes []byte) (err error) {
-	// flush if we are full
-	//
-	// XXX With OP_MSG the limit is larger; MaxBSONSize shouldn't be hard
-	// coded, it should be based on the server's ismaster response.
-	if bb.docCount >= bb.docLimit || bb.byteCount+len(rawBytes) > MaxBSONSize {
-		err = bb.Flush()
-	}
-	// buffer the document
+// Update adds a document to the buffer for bulk update. If the buffer becomes full, the bulk write is performed, returning
+// any error that occurs.
+func (bb *BufferedBulkInserter) Update(selector, update bson.D) error {
+	return bb.addModel(mongo.NewUpdateOneModel().SetFilter(selector).SetUpdate(update).SetUpsert(bb.upsert))
+}
+
+// Replace adds a document to the buffer for bulk replacement. If the buffer becomes full, the bulk write is performed, returning
+// any error that occurs.
+func (bb *BufferedBulkInserter) Replace(selector, replacement bson.D) error {
+	return bb.addModel(mongo.NewReplaceOneModel().SetFilter(selector).SetReplacement(replacement).SetUpsert(bb.upsert))
+}
+
+// InsertRaw adds a document, represented as raw bson bytes, to the buffer for bulk insertion. If the buffer becomes full,
+// the bulk write is performed, returning any error that occurs.
+func (bb *BufferedBulkInserter) InsertRaw(rawBytes []byte) error {
+	return bb.addModel(mongo.NewInsertOneModel().SetDocument(rawBytes))
+}
+
+func (bb *BufferedBulkInserter) addModel(model mongo.WriteModel) error {
```

Comment:
nit: while this is internal, this is where the interesting behavior described in docstrings above is actually happening. might be worth adding docstring here too.

Reply:
done